modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/PeriodicExpiryMessageDeletionTask.java

```diff
@@ -0,0 +1,158 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wso2.andes.kernel;
+
+import com.google.common.util.concurrent.SettableFuture;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.andes.kernel.slot.SafeDeletion;
+import org.wso2.andes.kernel.slot.SlotManagerClusterMode;
+import org.wso2.andes.kernel.slot.SlotManagerStandalone;
+import org.wso2.andes.store.FailureObservingStoreManager;
+import org.wso2.andes.store.HealthAwareStore;
+import org.wso2.andes.store.StoreHealthListener;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * This task is responsible to delete the expired messages from the database which were not currently
+ * allocated to any slots. If we delete any messages that are assigned to a slot delivery worker, the
+ * content may got deleted when the content is asked to retrieve. This can mess up the message delivery
+ * flow. Because of that the messages not allocated to a slot are considered as safe to delete
+ */
+public class PeriodicExpiryMessageDeletionTask implements Runnable, StoreHealthListener {
+
+    private static Log log = LogFactory.getLog(PeriodicExpiryMessageDeletionTask.class);
+
+    /**
+     * Indicates and provides a barrier if messages stores become offline.
+     * marked as volatile since this value could be set from a different thread
+     */
+    protected volatile SettableFuture<Boolean> messageStoresUnavailable;
+
+    /**
+     * Used to perform database operations on the context store.
+     */
+    private AndesContextStore andesContextStore = AndesContext.getInstance().getAndesContextStore();
+
+    /**
+     * Holds the slot manager based on broker running mode
+     */
+    private SafeDeletion safeDeletionSlotManager;
+
+    /**
+     * Indicate the cluster mode is enabled or not
+     */
+    private boolean isClusterEnabled;
+
+    public PeriodicExpiryMessageDeletionTask() {
+
+        this.messageStoresUnavailable = null;
+        // Register AndesRecoveryTask class as a StoreHealthListener
+        FailureObservingStoreManager.registerStoreHealthListener(this);
+
+        isClusterEnabled = AndesContext.getInstance().isClusteringEnabled();
+        // Set the appropriate slot manager
+        if (isClusterEnabled) {
+            this.safeDeletionSlotManager = SlotManagerClusterMode.getInstance();
```

Comment:
I think its wrong to do this check here. When the server start up I set the slot manager checking this condition. And I put a method to get the slot manager in that class also. So the correct way is to get it from there rather than checking this and set it here.